lineax/_solver/lsmr.py

```diff
@@ -0,0 +1,583 @@
+
+
+_LSMRState: TypeAlias = AbstractLinearOperator
+
+
+class LSMR(AbstractLinearSolver[_LSMRState], strict=True):
+    """LSMR solver for linear systems.
+
+    This solver can handle any operator, even nonsquare or singular ones. In these
+    cases it will return the pseudoinverse solution to the linear system.
+
+    Similar to `scipy.sparse.linalg.lsmr`.
+
+    This supports the following `options` (as passed to
+    `lx.linear_solve(..., options=...)`).
+
+    - `y0`: The initial estimate of the solution to the linear system. Defaults to all
+        zeros.
+    - `damp`: Damping factor for regularized least-squares. LSMR solves the regularized
+        least-squares problem:
+
+            min ||A y - b||_2 + damp*||y||_2
+
+        where damp is a scalar. If damp is None or 0, the system is solved without
+        regularization. Default is 0.
+
+    """
+
+    atol: float
+    btol: float
+    norm: Callable = two_norm
+    max_steps: Optional[int] = None
+    conlim: float = 1e8
+
+    def __check_init__(self):
+        if isinstance(self.atol, (int, float)) and self.atol < 0:
+            raise ValueError("Tolerances must be non-negative.")
+        if isinstance(self.btol, (int, float)) and self.btol < 0:
+            raise ValueError("Tolerances must be non-negative.")
+
+        if isinstance(self.atol, (int, float)) and isinstance(self.btol, (int, float)):
+            if self.atol == 0 and self.btol == 0 and self.max_steps is None:
+                raise ValueError(
+                    "Must specify `atol`, `btol`, or `max_steps` (or some combination "
+                    "of all three)."
+                )
+
+    def init(self, operator: AbstractLinearOperator, options: dict[str, Any]):
+        return operator
+
+    def compute(
+        self,
+        state: _LSMRState,
+        vector: PyTree[Array],
+        options: dict[str, Any],
+    ) -> tuple[PyTree[Array], RESULTS, dict[str, Any]]:
+        operator = state
+        x = options.get("y0", None)
+        damp = options.get("damp", 0.0)
+
+        m, n = operator.out_size(), operator.in_size()
+        # number of singular values
+        minDim = min([m, n])
+        if self.max_steps is None:
+            max_steps = minDim * 10  # for consistency with other iterative solvers
+        else:
+            max_steps = self.max_steps
+
+        if x is None:
+            x = jtu.tree_map(jnp.zeros_like, operator.in_structure())
+
+        b = vector
+        u = (ω(b) - ω(operator.mv(x))).ω
+        normb = self.norm(b)
+        beta = self.norm(u)
+
+        def beta_gt0(beta, u):
+            u = (ω(u) * (1 / beta)).ω
+            v = operator.T.mv(u)
+            alpha = self.norm(v)
+            return u, v, alpha
+
+        def beta_le0(beta, u):
+            v = jtu.tree_map(jnp.zeros_like, operator.in_structure())
+            alpha = 0.0
+            return u, v, alpha
+
+        u, v, alpha = lax.cond(beta > 0, beta_gt0, beta_le0, beta, u)
```

Comment:
Beta is a norm, how can be it be `<=0`, as then name suggests? (Is this check meant to specifically handle the `=0` case?)

Reply:
yeah this is really checking the beta==0 case, I'll rename things to make it clearer